patterns/strategy.md

```diff
@@ -0,0 +1,160 @@
+# Strategy (aka Policy)
+
+## Description
+
+The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design.
+It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
+The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem, we define only the skeleton of the algorithm at an abstract level and we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.
+
+In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
+In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
+This is why we call it "Dependency Inversion".
+
+## Motivation
+
+Imagine we are working on a project that generates reports every month.
+We need the reports to be generated in different formats (strategies), e.g.,
+in `JSON` or `Plain Text` formats.
+But things vary over time and we don't know what kind of requirement we may get in the future.
+For example, we may need to generate our report in a completly new format,
+or just modify one of the existing formats.
+
+## Example
+
+In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
+while `Text` and `Json` are our strategy structs.
+These strategies have to implement the `Formatter` trait.
+
+```rust
+use std::collections::HashMap;
+use std::fmt::Write as FmtWrite;
+use std::error::Error;
+type Data = HashMap<String, u32>;
+
+trait Formatter {
+    fn run(&self, data: &Data) -> Result<String, Box<dyn Error>>;
+}
+
+struct Report;
+
+impl Report {
+    fn generate<T: Formatter>(g: T) -> Result<String, Box<dyn Error>> {
+        // backend operations...
+        let mut data = HashMap::new();
+        data.insert("one".to_string(), 1);
+        data.insert("two".to_string(), 2);
+        // generate report
+        g.run(&data)
+    }
+}
+
+struct Text;
+impl Formatter for Text {
+    fn run(&self, data: &Data) -> Result<String, Box<dyn Error>> {
+        let mut s = String::new();
+
+        for (key, val) in data {
+            write!(&mut s, "{} {}\n", key, val)?;
```

Comment:
```suggestion
            let entry = format!("{} {}\n", key, val);
            s.push_str(&entry);
```
You use `write!` that may fail, but if you use something like this you are sure that it doesn't fail.
So basically all the `Result`s can be removed.

Reply:
This `Text`, in particular, could be written really elegantly with `map` I believe, but that's not required, it's just a thought I had.